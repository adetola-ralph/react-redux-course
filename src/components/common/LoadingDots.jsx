import React, { Component } from 'react';
import PropTypes from 'prop-types';

class LoadingDots extends Component {
  constructor(props, context) {
    super(props, context);

    this.state = {
      frame: 1,
    };
  }

  componentDidMount() {
    const { interval } = this.props;

    this.interval = setInterval(() => {
      this.setState({
        frame: this.state.frame + 1, // eslint-disable-line
      });
    }, interval);
  }

  componentWillUnmount() {
    clearInterval(this.interval);
  }

  render() {
    const stateDots = this.props.dots; // eslint-disable-line
    const { frame } = this.state;
    let dots = frame % (stateDots + 1);
    let text = '';

    while (dots > 0) {
      text += '.';
      dots -= 1;
    }

    return (
      <span {...this.props}>
        {text}
        &nbsp;
      </span>
    );
  }
}
LoadingDots.propTypes = {
  dots: PropTypes.number,
  interval: PropTypes.number,
};

LoadingDots.defaultProps = {
  dots: 3,
  interval: 300,
};

export default LoadingDots;
