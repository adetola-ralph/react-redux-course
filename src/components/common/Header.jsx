import React from 'react';
import { Link } from 'react-router-dom';
import PropTypes from 'prop-types';
import LoadingDots from './LoadingDots';

const Header = ({ loading }) => (
  <nav>
    <Link to="/">Home</Link>
    { '|' }
    <Link to="/about">About</Link>
    { '|' }
    <Link to="/courses">Courses</Link>
    { '|' }
    <Link to="/authors">Authors</Link>
    {loading && <LoadingDots interval={100} dots={20} />}
  </nav>
);

Header.propTypes = {
  loading: PropTypes.bool.isRequired,
};

export default Header;
