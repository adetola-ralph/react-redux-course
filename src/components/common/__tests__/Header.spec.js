import { create } from 'react-test-renderer';
import { MemoryRouter } from 'react-router-dom';

import { mount } from 'enzyme';

import Header from '../Header';
import LoadingDots from '../LoadingDots';

describe('Header component', () => {
  test('Snapshot', () => {
    const tree = create(
      <MemoryRouter>
        <Header loading/>
      </MemoryRouter>
    ).toJSON();
    expect(tree).toMatchSnapshot();
  });

  test('should contain loading dots component when loading is true', () => {
    const wrapper = mount(
      <MemoryRouter>
        <Header loading/>
      </MemoryRouter>
    );

    expect(wrapper.find(LoadingDots)).toHaveLength(1);
  });

  test('should contain loading dots component when loading is false', () => {
    const wrapper = mount(
      <MemoryRouter>
        <Header loading={false}/>
      </MemoryRouter>
    );

    expect(wrapper.find(LoadingDots)).toHaveLength(0);
  });
});
