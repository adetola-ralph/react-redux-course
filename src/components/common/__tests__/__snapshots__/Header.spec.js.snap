// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Header component Snapshot 1`] = `
<nav>
  <a
    href="/"
    onClick={[Function]}
  >
    Home
  </a>
  |
  <a
    href="/about"
    onClick={[Function]}
  >
    About
  </a>
  |
  <a
    href="/courses"
    onClick={[Function]}
  >
    Courses
  </a>
  |
  <a
    href="/authors"
    onClick={[Function]}
  >
    Authors
  </a>
  |
  <span>
    Number of courses: 
    0
  </span>
  <span>
    <span
      dots={20}
      interval={100}
    >
      .
       
    </span>
  </span>
</nav>
`;
